### PR TITLE
Allow creating repository from a template via the API

### LINKED ISSUE #11535: Allow creating repository from a template via the API
- Gitea version (or commit ref): 0b8b81e
- Git version:
- Operating system: arch
- Database (use `[x]`):
  - [x] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x] Not relevant
- Log gist:

## Description
Currently you can create a repository via the API, but not from a template, I know repo templates are a relatively new feature but it would be awesome if this could be included as a feature.

Sidenote: gitea is an awesome project by the way - big fan of this 🚀

### PR DESCRIPTION
**Pull request by 6543**
close #11535

### CODE REVIEW

**Comment by 6543** [comment]
I'm just thinking using a new api endpoint since extending `POST /user/repos` result in lot of fields who are never used in one either creating a new repo or from template 

perhaps `POST /user/repos/from_template` ?

**Comment by zeripath** [comment]
I think you should just be able to do this from the current API. What's the problem with just using "omitempty" on the Template settings bits if they're not being used?

File: modules/structs/repo.go [hunk 1]
```diff
@@ -116,6 +128,8 @@ type CreateRepoOption struct {
 	Readme string `json:"readme"`
 	// DefaultBranch of the repository (used when initializes and in template)
 	DefaultBranch string `json:"default_branch" binding:"GitRefName;MaxSize(100)"`
+	// AAAA
+	Template *CreateRepoTemplateOption
```
Comment (zeripath) [inline modules/structs/repo.go]:
```suggestion
	Template *CreateRepoTemplateOption `json:"template,omitempty" 
```

File: modules/structs/repo.go [hunk 2]
```diff
@@ -122,6 +134,8 @@ type CreateRepoOption struct {
 	// TrustModel of the repository
 	// enum: default,collaborator,committer,collaboratorcommitter
 	TrustModel string `json:"trust_model"`
+	// Template contain al options to create a repo based on a template
```
Comment (zeripath) [inline modules/structs/repo.go]:
```suggestion
	// Template contains all options to generate a repo based on a template
```

File: routers/api/v1/repo/repo.go [hunk 3]
```diff
@@ -234,19 +234,45 @@ func CreateUserRepo(ctx *context.APIContext, owner *models.User, opt api.CreateR
...
+
+	if opt.Template != nil {
+		templateRepo, err2 := models.GetRepositoryByID(opt.Template.RepoTemplate)
+		if err2 != nil {
+			if models.IsErrRepoNotExist(err) {
+				ctx.Error(http.StatusNotFound, "GetTemplateRepo", fmt.Errorf("Template Repo [%d] not found", opt.Template.RepoTemplate))
+			}
+			ctx.InternalServerError(err)
+			return
+		}
```
Comment (zeripath) [inline routers/api/v1/repo/repo.go:250]: Need to check if the templateRepository is a template

Comment (zeripath) [reply routers/api/v1/repo/repo.go:250]: still needs this check

File: modules/structs/repo.go [hunk 4]
```diff
@@ -93,6 +93,18 @@ type Repository struct {
 	Internal                  bool             `json:"internal"`
 }
 
+// CreateRepoOption options when creating repository
+// swagger:model
+type CreateRepoTemplateOption struct {
+	RepoTemplate int64
```
Comment (lunny) [inline modules/structs/repo.go:99]: Maybe we should use a repository full name but not a id?

**Comment by 6543** [comment]
->  #15958

**Pull request closed without merging**